scripts/automation/trex_control_plane/stl/trex_stl_lib/trex_stl_hltapi.py

```diff
@@ -424,15 +441,84 @@ def cleanup_session(self, **user_kwargs):
         self.trex_client = None
         return HLT_OK()
 
-    def interface_config(self, port_handle, mode='config'):
+    def interface_config(self, **user_kwargs):
         if not self.trex_client:
             return HLT_ERR('Connect first')
+
+        kwargs = merge_kwargs(interface_config_kwargs, user_kwargs)
+        is_array_based = not isinstance(kwargs['port_handle'], int)
+
+        mode = kwargs['mode']
         ALLOWED_MODES = ['config', 'modify', 'destroy']
         if mode not in ALLOWED_MODES:
             return HLT_ERR('Mode must be one of the following values: %s' % ALLOWED_MODES)
-        # pass this function for now...
-        return HLT_ERR('interface_config not implemented yet')
 
+        if is_array_based and not kwargs['port_handle']:
+            return HLT_ERR('Invalid port_handle parameter')
+
+        if not is_array_based:
+            kwargs['port_handle'] = [kwargs['port_handle']]
+            kwargs['intf_ip_addr'] = kwargs['intf_ip_addr'] and [kwargs['intf_ip_addr']]
+            kwargs['gateway'] = kwargs['gateway'] and [kwargs['gateway']]
+            kwargs['vlan'] = kwargs['vlan'] and [kwargs['vlan']]
+            kwargs['vlan_id'] = kwargs['vlan_id'] and [kwargs['vlan_id']]
+            kwargs['vlan_id_list'] = kwargs['vlan_id_list'] and [kwargs['vlan_id_list']]
+            kwargs['vlan_id_inner'] = kwargs['vlan_id_inner'] and [kwargs['vlan_id_inner']]
+
+        try:
+            self.trex_client.set_service_mode(kwargs['port_handle'], True)
+        except Exception as e:
+            return HLT_ERR('Error enabling service mode: %s' % e)
+
+        try:
+            if mode == 'destroy':
```

Comment:
Few notes regarding 'mode' variable.

1. AFAIK, mode 'destroy' should not implicitly continue to 'create' or 'modify'.
2. By spec, it should be list if port_handle is a list (with possible different values)
3. 'modify' and 'create' are same here

Reply:
@ybrustin, I have one fundamental question regarding "interfaces" in ixia, is there something corresponding that TRex has?

Here is the quote from ixia hltapi docs (mode parameter):
```
Action to be taken on the interface selected. This option takes a list of values
when -port_handle is a list of port handles. This option is valid for the old and
the new APIs. Valid options are:
   * config
   * destroy - clear all interfaces before configuring new ones.
   * modify - Valid only when using IxTclNetwork. When ::ixia::interface_config is
provided with -port_handle parameter, -mode modify and other supported parameters,
except -interface_handle, the modification is supported for L1 parameters only. When
::ixia::interface_config is provided with -port_handle parameter, -mode modify and
other supported parameters, including -interface_handle, the modification is supported
for L2-L3 parameters also, but only for protocol interfaces (-l23_config_type
protocol_interface). 
```
So, regarding list:
1. As I understand destroy means clearing all configs and create one from the scratch. But at TRex point, there is nothing to clear (or setting some default values), except vlan options.
2. Yep, my bad, I'll fix this.
3. After the first reading, I thought config and modify is almost the same for TRex, but for now, I'm not sure TRex can support `modify` at all. So maybe it would better to keep `config` and `destroy` and make it returns `HLT_ERR` on `modify` mode?

Reply:
> As I understand destroy means clearing all configs and create one from the scratch. But at TRex point, there is nothing to clear (or setting some default values), except vlan options.

I hoped you did some research regarding what should be the behavior.
Let's ask (internally) some experts that might have answers.
Did you check that it works with existing HLT tests?

> After the first reading, I thought config and modify is almost the same for TRex, but for now, I'm not sure TRex can support modify at all. So maybe it would better to keep config and destroy and make it returns HLT_ERR on modify mode?

That was I would do, better do declare something is not supported, than declare it's supported and behavior will be incorrect...